### ISSUE TITLE
calicoctl node can fail to detect successful docker image startup

### ISSUE THREAD

**Issue by DreadPirateShawn** [projectcalico/calico-containers#1066]
We occasionally hit a scenario where "calicoctl node" appears to hang, but we can see that the calico-node image has in fact started successfully.

We run `calicoctl node --ip=192.168.0.68 --detach=true` and we see the Calico process start just fine, we see the "calico-node" appear in "docker ps", but the `calicoctl node` process itself never completes.

Calico processes:

```
$ ps aux | grep calico
highland  2693  0.0  0.0   4408   608 pts/1    Ss+  08:03   0:00 sh -c sudo /opt/skytap/bin/skytap-service calico start
root      2694  0.0  0.1 109824  3172 pts/1    S+   08:03   0:00 sudo /opt/skytap/bin/skytap-service calico start
root      2695  0.0  0.0  12588  1796 pts/1    S+   08:03   0:00 /bin/bash /etc/init.d/calicoctl start
root      2870  0.0  0.0   9716  1788 pts/1    S+   08:03   0:00 /opt/calico/bin/calicoctl node --ip=192.168.0.68 --detach=true
root      2872  0.0  0.8 107076 17524 pts/1    S+   08:03   0:00 /opt/calico/bin/calicoctl node --ip=192.168.0.68 --detach=true
root      2948  0.0  0.0    748     4 ?        S    08:03   0:00 svlogd -tt /var/log/calico/bird6
root      2949  0.0  0.0    640   252 ?        S    08:03   0:05 bird6 -R -s /var/run/calico/bird6.ctl -d -c /etc/calico/confd/config/bird6.cfg
root      2951  0.0  0.0    752     4 ?        S    08:03   0:00 svlogd /var/log/calico/felix
root      2952  0.0  0.0    748     4 ?        S    08:03   0:00 svlogd -tt /var/log/calico/bird
root      2953  0.0  0.0    748     4 ?        S    08:03   0:00 svlogd /var/log/calico/confd
root      2956  0.0  0.0    716   516 ?        S    08:03   0:07 bird -R -s /var/run/calico/bird.ctl -d -c /etc/calico/confd/config/bird.cfg
root      2957  0.1  1.2 115028 24656 ?        Sl   08:03   0:53 /usr/bin/python /usr/bin/calico-felix
root      2959  0.0  0.2  10692  4596 ?        Sl   08:03   0:00 confd -confdir=/etc/calico/confd -interval=5 -watch --log-level=debug -node=http://127.0.0.1:4001 -client-key= -client-cert= -client-ca-keys=
root      2978  0.3  1.1 115824 23008 ?        Sl   08:03   2:32 /usr/bin/python -m calico.etcddriver /run/felix-driver.sck
```

"calico-node" logs:

```
$ docker logs calico-node
Calico node started successfully
```

Unfortunately we were tee-ing the output to a temp file, which never flushed to disk, so the raw stdout is not available from this repro -- but given the processes and Docker image, we believe it was:

```
Running Docker container with the following command:

docker run -d --restart=always --net=host --privileged --name=calico-node -e HOSTNAME=r1knode3 -e IP=192.168.0.68 -e IP6= -e CALICO_NETWORKING=true -e AS= -e NO_DEFAULT_POOLS=TRUE -e ETCD_AUTHORITY=127.0.0.1:4001 -e ETCD_SCHEME=http -v /highland/logs/calico:/var/log/calico -v /var/run/calico:/var/run/calico calico/node:v0.18.0

Calico node is running with id: 1d240e9b93d4486ec1a500051a25cc0beccc96a3714cbe2a32bdd488b4d69368
Waiting for successful startup
```

We _think_ that the problem might be this:

https://github.com/projectcalico/calico-containers/blob/master/calicoctl/calico_ctl/node.py

```
        for raw_data in output:
            sys.stdout.write(raw_data)
            if "Calico node started successfully" in raw_data and startup_only:
                stop_container_on_exit = False
                break
```

...which can misbehave if the output stream is chunked such that the string "Calico node started successfully" is not sent all at once, and would presumably manifest in exactly the repro that we see.

**Comment by fasaxc** [comment]
@DreadPirateShawn when you say "occasionally", do you have any feel for how often (across how many systems) and if there are any particular circumstances (e.g. system under load)?

That loop feels like a good thread to pull on.  Normally doing `for x in file_like_object:` does iterate over lines but I could believe that we're hitting a buffer-flushing issue or that the `output` object isn't file-like enough.

**Comment by DreadPirateShawn** [comment]
@fasaxc By "occasionally", we've seen it twice so far, almost a month apart, during nightly automated "create VMs and spin up all services sequentially" tests, of which we run about 2-3 per night. Granted, the tests don't always reach the Calico startup phase, so the numbers aren't clear-cut, but rough estimates put it at less than 1% of Calico node startups. Definitely rare, but on the other hand "twice in one month in an enterprise environment."

As for load, we start our services sequentially -- at the point of this repro, very little has been started on the host. Post-reboot, so upstart launches Docker, then we start etcd proxy, then Calico node (and this repro).

**Comment by fasaxc** [comment]
@DreadPirateShawn I can't reproduce the issue locally but, based on your hunch, here's a patched calicoctl binary that will output each chunk it reads from docker as well as concatenating the output before scanning it for the marker string.

```
diff --git a/calicoctl/calico_ctl/node.py b/calicoctl/calico_ctl/node.py
index fb671f9..4290bd0 100644
--- a/calicoctl/calico_ctl/node.py
+++ b/calicoctl/calico_ctl/node.py
@@ -839,10 +839,12 @@ def _attach_and_stream(container, startup_only):
     exit_code = 1
 
     output = docker_client.attach(container, stream=True)
+    all_output = ""
     try:
         for raw_data in output:
-            sys.stdout.write(raw_data)
-            if "Calico node started successfully" in raw_data and startup_only:
+            sys.stdout.write("DOCKER OUTPUT: %r\n" % raw_data)
+            all_output += raw_data
+            if "Calico node started successfully" in all_output and startup_only:
                 stop_container_on_exit = False
                 break
     except KeyboardInterrupt:
```

```
# ./calicoctl node --detach=true
Running Docker container with the following command:

docker run -d --restart=always --net=host --privileged --name=calico-node -e HOSTNAME=smc-ubuntu -e IP= -e IP6= -e CALICO_NETWORKING=true -e AS= -e NO_DEFAULT_POOLS= -e ETCD_AUTHORITY=127.0.0.1:2379 -e ETCD_SCHEME=http -v /var/log/calico:/var/log/calico -v /lib/modules:/lib/modules -v /var/run/calico:/var/run/calico calico/node:v0.20.0

Calico node is running with id: 763be4259e7cd513c05436b61c8381328f75ccf5fb739f1e74bce19926d74e50
Waiting for successful startup
DOCKER OUTPUT: 'No IP provided. Using detected IP: 172.16.171.5\n'
DOCKER OUTPUT: 'Calico node started successfully\n'
```

Please can you give that a spin and see if you still hit the issue (and if you do hit the issue, please can you grab the output from calicoctl)?

[calicoctl-read-patch.zip](https://github.com/projectcalico/calico-containers/files/385676/calicoctl-read-patch.zip)

**Comment by DreadPirateShawn** [comment]
Sorry for the delay -- will do, I plan to get a custom repro loop with the attached binary set up later today.

**Comment by DreadPirateShawn** [comment]
Well, I've had this custom binary starting & stopping in a loop for over a week now, in parallel with the legacy binary -- *none* of them have manifested the problem so far.

That being said, we can't really prove a negative, of course. It could be that starting & stopping the calico node itself isn't a sufficient repro, it may also only occur when a host has just booted up and docker service has just started -- which is a heavier repro scenario than I have currently automated.

So -- one option here is to proceed with pushing this likely fix, resolve this ticket, and we'll upgrade and see if the original problem ever repros in the newer version.

(If you'd rather push for a heavier repro first, let me know -- but figured I'd sync first to see what plan you'd prefer.)

**Comment by fasaxc** [comment]
I'm happy to make a tentative fix here rather than do a dragged out repro.  That patch isn't production ready, though so we'll need to respin to avoid infinitely concatenating to the `all_output` string.

**Comment by caseydavenport** [comment]
@fasaxc this guy is pretty old - what's the latest on this?

**Comment by fasaxc** [comment]
We merged a tentative fix.  I'm going to close.  @DreadPirateShawn please raise a new issue if you see this again after the above fix went in.
